### PR TITLE
[Live] Control the type of generated URLs

### PR DESCRIPTION
**Pull request by 1ed**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | -
| License       | MIT

Add a way to generate absolute URLs for live components.

My use case is that I embed a page with live components 
into a site on a different host, so I need absolute URLs 
for the components to work properly.

### CODE REVIEW

**Comment by smnandre** [comment]
Seems to me a very specific use case.. could it be handled with router configuration ?

File: src/LiveComponent/src/Attribute/AsLiveComponent.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 namespace Symfony\UX\LiveComponent\Attribute;
 
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
```
Comment (smnandre) [inline src/LiveComponent/src/Attribute/AsLiveComponent.php:14]: Not sure we should couple UrlGeneratorInterface and this attribute.

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php [hunk 2]
```diff
@@ -89,7 +89,7 @@ public function load(array $configs, ContainerBuilder $container): void
             AsLiveComponent::class,
             function (ChildDefinition $definition, AsLiveComponent $attribute) {
                 $definition
-                    ->addTag('twig.component', array_filter($attribute->serviceConfig()))
+                    ->addTag('twig.component', array_filter($attribute->serviceConfig(), static fn ($v) => null !== $v))
```
Comment (smnandre) [inline src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: Is this related to this PR ?

Comment (1ed) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: Without this, it does not work. It's a separate commit, so it can be cherry-picked as a bugfix or it can be moved to its own PR if needed.

File: src/LiveComponent/src/Attribute/AsLiveComponent.php (hunk 1 above)
Comment (1ed) [reply src/LiveComponent/src/Attribute/AsLiveComponent.php:14]: We can invent or own constants and map them over, but the component already bound with the routing component.

**Review by 1ed — COMMENTED** [review]
> Seems to me a very specific use case.. could it be handled with router configuration ?

IDK maybe.. what do you have in mind?

If you suggesting to use the `host` parameter it doesn't work, because from Symfony's point of view the host is the same as the request comes in to.

File: src/LiveComponent/src/Attribute/AsLiveComponent.php (hunk 1 above)
Comment (weaverryan) [reply src/LiveComponent/src/Attribute/AsLiveComponent.php:14]: I agree - we're already bound to routing - so I'm ok with it

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (weaverryan) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: Is there an error without it? I actually can't remember why we have that `array_filter()` at all

**Review by weaverryan — COMMENTED** [review]
This looks sensible to me!

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (1ed) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: If I remove the `array_filter` it breaks the compiler pass in TwigComponent...

https://github.com/symfony/ux/blob/a2418ef146f09a858bd12ed0116c3f8058d0a3a8/src/TwigComponent/src/DependencyInjection/Compiler/TwigComponentPass.php#L48

and other places where `null` is not checked.

File: src/LiveComponent/src/Attribute/AsLiveComponent.php (hunk 1 above)
Comment (smnandre) [reply src/LiveComponent/src/Attribute/AsLiveComponent.php:14]: Allright then

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (smnandre) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: Yes but would that break things now that you allow empty strings ?

**Comment by smnandre** [comment]
Wait, so you'll POST from domainA.com to domainB.com ? 

Should this be on a component - per - component basis... or a config wide one (that seem more logical to me) ?

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (1ed) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: I think it's not, all tests are green, but I can add `'' !== $v` too, if needed.

**Comment by 1ed** [comment]
> Wait, so you'll POST from domainA.com to domainB.com ?
> 
> Should this be on a component - per - component basis... or a config wide one (that seem more logical to me) ?

As the route can be configured on the component level, for me it is logical to do the same with the URL generated for that route.

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (1ed) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: OK, I added back filtering out empty strings too.

**Comment by smnandre** [comment]
> As the route can be configured on the component level, for me it is logical to do the same with the URL generated for that route.

Yes i mean, should this be possible to config it on an app-level too, as the route can be configured there for all components ? 

Last thing, is there any implication/effect with the "liveQuery" thing ?

**Comment by 1ed** [comment]
> > As the route can be configured on the component level, for me it is logical to do the same with the URL generated for that route.
> 
> Yes i mean, should this be possible to config it on an app-level too, as the route can be configured there for all components ?

Ah, I see. That would mean we need to introduce configuration for the live component and add a config entry for it, e.g. something like

```yaml
live_component:
    default_url_reference_type: 'absolute_path' # support strings and UrlGeneratorInterface consts too -> map strings back to consts
```

read the configuration and use this as the default value for the metadata (the default in `AsLiveComponent` should be null) e.g. in `LiveComponentMetadataFactory`. That can be done, maybe in a different PR.
 
> Last thing, is there any implication/effect with the "liveQuery" thing ?

I'm not sure, I'm not familiar with that, but I took a look at the PR and I think using the browser's history api and getting query string values from the request should not be affected by this.

File: src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php (hunk 2 above)
Comment (smnandre) [reply src/LiveComponent/src/DependencyInjection/LiveComponentExtension.php]: Man it took me so much time to understand i'm sorry .. 

I suggest that we handle that in another PR, as it's clearly not in LiveComponent that we should handle this.

https://github.com/symfony/ux/pull/1402

**Comment by smnandre** [comment]
Could you add a test with an action (just to be sure everything work with in JS and in the request handling ? 

(and that way we'd have a regression test later on this)

**Comment by 1ed** [comment]
> Could you add a test with an action (just to be sure everything work with in JS and in the request handling ?
> 
> (and that way we'd have a regression test later on this)

I just added a test for that (but without js, for that we would need to setup panther, I think).

**Comment by 1ed** [comment]
I think this is ready. Can we merge it in this form or it needs more work?

**Review by weaverryan — APPROVED** [review]

**Comment by weaverryan** [comment]
Thanks Gábor!
